rocmPackages.composable_kernel: compress output

mschwaig opened the pull request:
## Description of changes

I made a mistake in my ROCm 6.0 PR https://github.com/NixOS/nixpkgs/pull/287846, not realizing that `rocmPackages.composable_kernel` would be larger than Hydra's size limit. See the comments here for details: https://github.com/NixOS/nixpkgs/pull/287846#issuecomment-2016780211

~This PR compresses the huge `lib/libdevice_operations.a` file that's in the output of that derivation, and then decompresses it again in another derivation based on `runCommandLocal`.~

~Is this kind of fix acceptable?~
~Is how I implemented this configurable enough and the derivations carry the correct metadata?~
~How else could this issue be addressed?~
~Not having that derivation cached is a huge pain, since the build takes a few hours at least.~

~I am running `nixpkgs-review` on this right now.~

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

mschwaig commented:
Result of `nixpkgs-review pr 299589` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>rocmPackages.migraphx</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>rocmPackages.composable_kernel</li>
    <li>rocmPackages.miopen</li>
  </ul>
</details>

ulrikstrid commented:
I'm going to invoke @wegank and @SuperSandro2000 to give their view of this

SuperSandro2000 commented:
> `lib/libdevice_operations.a`

Do we even need the static library? Is it stripped?

GZGavinZhao commented:
> Do we even need the static library?

It is intended to be consumed as a static library. Building it as a shared library needs patching. I've tried that in the past, and every time MIOpen would fail to find composable_kernel properly if I do that. I will try this again and see if the situation now improves.

GZGavinZhao commented:
An option would be to cherry-pick https://github.com/llvm/llvm-project/commit/7e2823438e920d25364ff92b62ad90020c31bb59 to ROCm's LLVM.

Or maybe backport https://github.com/ROCm/composable_kernel/issues/1053 and https://github.com/ROCm/MIOpen/pull/2526. I've tried this a few months ago (https://github.com/GZGavinZhao/composable_kernel/commit/c8d9ac8eb68f710a41dc9151cd5e463febd49833) and I don't recall getting too much space improvement, but I could've mis-remembered since my original intention wasn't to save space.

mschwaig commented:
> > `lib/libdevice_operations.a`
> 
> Do we even need the static library? Is it stripped?

Yes, we need it. It is a fat binary file with a a lot of of GPU-specific code for each target GPU, which is used by other parts of ROCm.

I checked the log and yes, it is stripped:
```
stripping (with command strip and flags -S -p) in  /nix/store/sfybgqj64vhgp386fd3xw0fkjx0wvzwf-composable_kernel-6.0.2/lib
```
I would prefer if we first found an acceptable solution to get caching to work for users of ROCm, because especially `composable_kernel` is a real pain to build in terms of hardware requirements and how long the build takes. We could then always switch to a better solution when we find one that we can implement with a reasonable amount of effort.

Another alternative to this PR would probably be setting a long enough `meta.timeout` for the few consumers of `composable_kernel` (`migraphx` and `miopen`) so it can be rebuilt on the fly while building those.
Though I barely know anything about Hydra, so I don't know if it would behave how I would imagine.

In case the bandwidth and storage costs are actually proportional to compressed size I think it would make sense to tie Hydra's output size limit to compressed size to avoid these kinds of workarounds as much as possible.

mschwaig commented:
We could also try dropping the gfx940 and gfx941 targets as suggested here https://github.com/NixOS/nixpkgs/pull/298388#issuecomment-2032791130 to see if that gets us below the size limit.

mschwaig commented:
Here is a bit of an overview of what's in that file:

```
$ bloaty /nix/store/sfybgqj64vhgp386fd3xw0fkjx0wvzwf-composable_kernel-6.0.2/lib/libdevice_operations.a
    FILE SIZE        VM SIZE    
 --------------  -------------- 
  75.4%  2.78Gi  93.7%  2.78Gi    .hip_fatbin
   8.1%   307Mi   0.0%       0    .strtab
   5.9%   221Mi   0.0%       0    [AR Symbol Table]
   5.8%   219Mi   4.7%   142Mi    [954581 Others]
   2.4%  90.7Mi   0.0%       0    [ELF Section Headers]
   0.8%  29.9Mi   1.0%  29.9Mi    .rodata.str1.1
   0.6%  23.6Mi   0.0%       0    .symtab
   0.4%  13.6Mi   0.4%  13.6Mi    .eh_frame
   0.2%  8.49Mi   0.0%       0    .rela.eh_frame
   0.1%  5.64Mi   0.0%       0    .group
   0.1%  3.86Mi   0.1%  3.86Mi    .text
   0.1%  2.97Mi   0.0%       0    .rela.text
   0.0%   877Ki   0.0%       0    [Unmapped]
   0.0%   688Ki   0.0%       0    .rela.text._ZN2ck15get_device_nameB5cxx11Ev
   0.0%   522Ki   0.0%   522Ki    .text._ZN2ck15get_device_nameB5cxx11Ev
   0.0%   361Ki   0.0%       0    .llvm_addrsig
   0.0%   310Ki   0.0%       0    .rela.text._Z15hip_check_error10hipError_t
   0.0%   165Ki   0.0%       0    .rela.text._ZNK2ck16tensor_operation6device12BaseOperator17GetTypeIdHashCodeB5cxx11Ev
   0.0%   160Ki   0.0%   160Ki    .text._Z15hip_check_error10hipError_t
   0.0%   153Ki   0.0%   153Ki    .text._ZNK2ck16tensor_operation6device12BaseOperator17GetTypeIdHashCodeB5cxx11Ev
   0.0%   105Ki   0.0%       0    .rela.data.rel.ro
 100.0%  3.69Gi 100.0%  2.97Gi    TOTAL
```

`.hip_fatbin` contains the GPU specific code and this is what `clang-offload-bundler` will be able to compress in the future.

Flakebi commented:
The workaround here looks good to me.
(Much more appealing than rebuilding for its dependencies.)

Thanks for putting so much time and effort into this.

mschwaig commented:
> Another alternative to this PR would probably be setting a long enough `meta.timeout` for the few consumers of `composable_kernel` (`migraphx` and `miopen`) so it can be rebuilt on the fly while building those.
> Though I barely know anything about Hydra, so I don't know if it would behave how I would imagine.

Turns out this would not even work because `miopen` is practically the same size as `composable_kernel` so it will not get cached either. In practice this PR still reduces the problem by orders of magnitude, since building `miopen` will only take ROCm users about 10 minutes. Packages further downstream like `torchWithRocm` should be cached, since they are a lot smaller.

mschwaig commented:
> We could also try dropping the gfx940 and gfx941 targets as suggested here [#298388 (comment)](https://github.com/NixOS/nixpkgs/pull/298388#issuecomment-2032791130) to see if that gets us below the size limit.

I tried this as well now and it does not work either. It only gets us down to 3.1 GB.

mschwaig commented:
> An option would be to cherry-pick [llvm/llvm-project@7e28234](https://github.com/llvm/llvm-project/commit/7e2823438e920d25364ff92b62ad90020c31bb59) to ROCm's LLVM.

I have backported this commit here https://github.com/mschwaig/llvm-project/commit/527b6eaf13b5bdcf3d825d150badb091b330e9f4 and it looks like that will work. The size of `composable_kernel` and `miopen` goes down from 3.7 GB to 1.3 GB when I do this.

What I'm still having trouble with:
1. Applying this commit as a patch in Nix. For some reason none of the changes, besides the ones creating new files, apply at all because the files in question are not found. I suspect that this has something to do with `sourceRoot`, but I could not figure out how to get past that issue.
2. Right now I'm just wrapping `clang-offload-bundler` to add the `-compress` argument instead of only modifying the relevant invocations, which I was not able to identify yet. This might be fine, but I am not sure.
3. I have not tested that applications keep working.

You can find what I have here: https://github.com/mschwaig/nixpkgs/commit/fe795d453cc298ea436bd27d6896fdd9d2ec8aa6

To conserve reviewer's time I am moving this back into the draft stage.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/trouble-patching-llvm-source-code-for-rocm/42772/1

mschwaig committed:
rocmPackages.llvm: replace --replace with --replace-fail (cleanup)

mschwaig commented:
Result of `nixpkgs-review pr 299589` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>22 packages marked as broken and skipped:</summary>
  <ul>
    <li>rocmPackages.llvm.flang</li>
    <li>rocmPackages.llvm.flang.doc</li>
    <li>rocmPackages.llvm.flang.info</li>
    <li>rocmPackages.llvm.flang.man</li>
    <li>rocmPackages.llvm.libclc</li>
    <li>rocmPackages.migraphx</li>
    <li>rocmPackages.mivisionx</li>
    <li>rocmPackages.mivisionx-cpu</li>
    <li>rocmPackages.mivisionx-hip</li>
    <li>rocmPackages.rdc</li>
    <li>rocmPackages.rdc.doc</li>
    <li>rocmPackages_6.llvm.flang</li>
    <li>rocmPackages_6.llvm.flang.doc</li>
    <li>rocmPackages_6.llvm.flang.info</li>
    <li>rocmPackages_6.llvm.flang.man</li>
    <li>rocmPackages_6.llvm.libclc</li>
    <li>rocmPackages_6.migraphx</li>
    <li>rocmPackages_6.mivisionx</li>
    <li>rocmPackages_6.mivisionx-cpu</li>
    <li>rocmPackages_6.mivisionx-hip</li>
    <li>rocmPackages_6.rdc</li>
    <li>rocmPackages_6.rdc.doc</li>
  </ul>
</details>
<details>
  <summary>88 packages built:</summary>
  <ul>
    <li>blender-hip</li>
    <li>rocmPackages.clang-ocl (rocmPackages_6.clang-ocl)</li>
    <li>rocmPackages.clr (rocmPackages_6.clr)</li>
    <li>rocmPackages.clr.icd (rocmPackages_6.clr.icd)</li>
    <li>rocmPackages.composable_kernel (rocmPackages_6.composable_kernel)</li>
    <li>rocmPackages.half (rocmPackages_6.half)</li>
    <li>rocmPackages.hip-common (rocmPackages_6.hip-common)</li>
    <li>rocmPackages.hipblas (rocmPackages_6.hipblas)</li>
    <li>rocmPackages.hipcc (rocmPackages_6.hipcc)</li>
    <li>rocmPackages.hipcub (rocmPackages_6.hipcub)</li>
    <li>rocmPackages.hipfft (rocmPackages_6.hipfft)</li>
    <li>rocmPackages.hipfort (rocmPackages_6.hipfort)</li>
    <li>rocmPackages.hipify (rocmPackages_6.hipify)</li>
    <li>rocmPackages.hiprand (rocmPackages.rocrand ,rocmPackages_6.hiprand ,rocmPackages_6.rocrand)</li>
    <li>rocmPackages.hipsolver (rocmPackages_6.hipsolver)</li>
    <li>rocmPackages.hipsparse (rocmPackages_6.hipsparse)</li>
    <li>rocmPackages.hsa-amd-aqlprofile-bin (rocmPackages_6.hsa-amd-aqlprofile-bin)</li>
    <li>rocmPackages.llvm.bintools (rocmPackages_6.llvm.bintools)</li>
    <li>rocmPackages.llvm.clang (rocmPackages_6.llvm.clang)</li>
    <li>rocmPackages.llvm.clang-tools-extra (rocmPackages_6.llvm.clang-tools-extra)</li>
    <li>rocmPackages.llvm.clang-tools-extra.doc (rocmPackages_6.llvm.clang-tools-extra.doc)</li>
    <li>rocmPackages.llvm.clang-tools-extra.info (rocmPackages_6.llvm.clang-tools-extra.info)</li>
    <li>rocmPackages.llvm.clang-tools-extra.man (rocmPackages_6.llvm.clang-tools-extra.man)</li>
    <li>rocmPackages.llvm.clang-unwrapped (rocmPackages_6.llvm.clang-unwrapped)</li>
    <li>rocmPackages.llvm.clang-unwrapped.doc (rocmPackages_6.llvm.clang-unwrapped.doc)</li>
    <li>rocmPackages.llvm.clang-unwrapped.info (rocmPackages_6.llvm.clang-unwrapped.info)</li>
    <li>rocmPackages.llvm.clang-unwrapped.man (rocmPackages_6.llvm.clang-unwrapped.man)</li>
    <li>rocmPackages.llvm.compiler-rt (rocmPackages_6.llvm.compiler-rt)</li>
    <li>rocmPackages.llvm.libc (rocmPackages_6.llvm.libc)</li>
    <li>rocmPackages.llvm.libc.doc (rocmPackages_6.llvm.libc.doc)</li>
    <li>rocmPackages.llvm.libcxx (rocmPackages_6.llvm.libcxx)</li>
    <li>rocmPackages.llvm.libcxx.doc (rocmPackages_6.llvm.libcxx.doc)</li>
    <li>rocmPackages.llvm.libcxxabi (rocmPackages_6.llvm.libcxxabi)</li>
    <li>rocmPackages.llvm.libunwind (rocmPackages_6.llvm.libunwind)</li>
    <li>rocmPackages.llvm.libunwind.doc (rocmPackages_6.llvm.libunwind.doc)</li>
    <li>rocmPackages.llvm.lld (rocmPackages_6.llvm.lld)</li>
    <li>rocmPackages.llvm.lld.doc (rocmPackages_6.llvm.lld.doc)</li>
    <li>rocmPackages.llvm.lldb (rocmPackages_6.llvm.lldb)</li>
    <li>rocmPackages.llvm.lldb.doc (rocmPackages_6.llvm.lldb.doc)</li>
    <li>rocmPackages.llvm.lldb.info (rocmPackages_6.llvm.lldb.info)</li>
    <li>rocmPackages.llvm.lldb.man (rocmPackages_6.llvm.lldb.man)</li>
    <li>rocmPackages.llvm.llvm (rocmPackages_6.llvm.llvm)</li>
    <li>rocmPackages.llvm.llvm.doc (rocmPackages_6.llvm.llvm.doc)</li>
    <li>rocmPackages.llvm.llvm.info (rocmPackages_6.llvm.llvm.info)</li>
    <li>rocmPackages.llvm.llvm.man (rocmPackages_6.llvm.llvm.man)</li>
    <li>rocmPackages.llvm.mlir (rocmPackages_6.llvm.mlir)</li>
    <li>rocmPackages.llvm.openmp (rocmPackages_6.llvm.openmp)</li>
    <li>rocmPackages.llvm.openmp.doc (rocmPackages_6.llvm.openmp.doc)</li>
    <li>rocmPackages.llvm.openmp.info (rocmPackages_6.llvm.openmp.info)</li>
    <li>rocmPackages.llvm.openmp.man (rocmPackages_6.llvm.openmp.man)</li>
    <li>rocmPackages.llvm.polly (rocmPackages_6.llvm.polly)</li>
    <li>rocmPackages.llvm.polly.doc (rocmPackages_6.llvm.polly.doc)</li>
    <li>rocmPackages.llvm.polly.info (rocmPackages_6.llvm.polly.info)</li>
    <li>rocmPackages.llvm.polly.man (rocmPackages_6.llvm.polly.man)</li>
    <li>rocmPackages.llvm.pstl (rocmPackages_6.llvm.pstl)</li>
    <li>rocmPackages.llvm.rocmClangStdenv (rocmPackages_6.llvm.rocmClangStdenv)</li>
    <li>rocmPackages.miopen (rocmPackages.miopen-hip ,rocmPackages_6.miopen ,rocmPackages_6.miopen-hip)</li>
    <li>rocmPackages.rccl (rocmPackages_6.rccl)</li>
    <li>rocmPackages.rocalution (rocmPackages_6.rocalution)</li>
    <li>rocmPackages.rocblas (rocmPackages_6.rocblas)</li>
    <li>rocmPackages.rocdbgapi (rocmPackages_6.rocdbgapi)</li>
    <li>rocmPackages.rocdbgapi.doc (rocmPackages_6.rocdbgapi.doc)</li>
    <li>rocmPackages.rocfft (rocmPackages_6.rocfft)</li>
    <li>rocmPackages.rocgdb (rocmPackages_6.rocgdb)</li>
    <li>rocmPackages.rocm-cmake (rocmPackages_6.rocm-cmake)</li>
    <li>rocmPackages.rocm-comgr (rocmPackages_6.rocm-comgr)</li>
    <li>rocmPackages.rocm-core (rocmPackages_6.rocm-core)</li>
    <li>rocmPackages.rocm-device-libs (rocmPackages_6.rocm-device-libs)</li>
    <li>rocmPackages.rocm-runtime (rocmPackages_6.rocm-runtime)</li>
    <li>rocmPackages.rocm-smi (rocmPackages_6.rocm-smi)</li>
    <li>rocmPackages.rocm-thunk (rocmPackages_6.rocm-thunk)</li>
    <li>rocmPackages.rocminfo (rocmPackages_6.rocminfo)</li>
    <li>rocmPackages.rocmlir (rocmPackages_6.rocmlir)</li>
    <li>rocmPackages.rocmlir-rock (rocmPackages_6.rocmlir-rock)</li>
    <li>rocmPackages.rocmlir.external (rocmPackages_6.rocmlir.external)</li>
    <li>rocmPackages.rocprim (rocmPackages_6.rocprim)</li>
    <li>rocmPackages.rocprofiler (rocmPackages_6.rocprofiler)</li>
    <li>rocmPackages.rocr-debug-agent (rocmPackages_6.rocr-debug-agent)</li>
    <li>rocmPackages.rocsolver (rocmPackages_6.rocsolver)</li>
    <li>rocmPackages.rocsparse (rocmPackages_6.rocsparse)</li>
    <li>rocmPackages.rocthrust (rocmPackages_6.rocthrust)</li>
    <li>rocmPackages.roctracer (rocmPackages_6.roctracer)</li>
    <li>rocmPackages.rocwmma (rocmPackages_6.rocwmma)</li>
    <li>rocmPackages.rpp (rocmPackages.rpp-hip ,rocmPackages_6.rpp ,rocmPackages_6.rpp-hip)</li>
    <li>rocmPackages.rpp-cpu (rocmPackages_6.rpp-cpu)</li>
    <li>rocmPackages.rpp-opencl (rocmPackages_6.rpp-opencl)</li>
    <li>rocmPackages.tensile (rocmPackages_6.tensile)</li>
    <li>rocmPackages.tensile.dist (rocmPackages_6.tensile.dist)</li>
  </ul>
</details>

mschwaig committed:
rocmPackages.llvm: compress outputs of clang-offload-bundler

This patches the clang-offload-bundler tool to add a compression option
from a more recent version of clang.
This compression option reduces the size of ROCm's fat binaries.
Those binaries contain .hip_fatbin sections with GPU-specific code,
for each target.
Compression is automatically turned on for all produced outputs
via a wrapper, because it's difficult to identify all the places
where the -compression argument would be needed.

Once upsteam introduces handeling for this argument,
we should drop the wrapper again.
This transistion will create inconsistsency, but I do not think that
it will impact any actual users and it's what's practical to implement.

mschwaig commented:
I have a working version with the `-compress` option for `clang-offload-bundler` now, so this is ready for review again.

Result of `nixpkgs-review pr 299589` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>11 packages marked as broken and skipped:</summary>
  <ul>
    <li>rocmPackages.llvm.flang</li>
    <li>rocmPackages.llvm.flang.doc</li>
    <li>rocmPackages.llvm.flang.info</li>
    <li>rocmPackages.llvm.flang.man</li>
    <li>rocmPackages.llvm.libclc</li>
    <li>rocmPackages.migraphx</li>
    <li>rocmPackages.mivisionx</li>
    <li>rocmPackages.mivisionx-cpu</li>
    <li>rocmPackages.mivisionx-hip</li>
    <li>rocmPackages.rdc</li>
    <li>rocmPackages.rdc.doc</li>
  </ul>
</details>
<details>
  <summary>88 packages built:</summary>
  <ul>
    <li>blender-hip</li>
    <li>rocmPackages.clang-ocl</li>
    <li>rocmPackages.clr</li>
    <li>rocmPackages.clr.icd</li>
    <li>rocmPackages.composable_kernel</li>
    <li>rocmPackages.half</li>
    <li>rocmPackages.hip-common</li>
    <li>rocmPackages.hipblas</li>
    <li>rocmPackages.hipcc</li>
    <li>rocmPackages.hipcub</li>
    <li>rocmPackages.hipfft</li>
    <li>rocmPackages.hipfort</li>
    <li>rocmPackages.hipify</li>
    <li>rocmPackages.hiprand</li>
    <li>rocmPackages.hipsolver</li>
    <li>rocmPackages.hipsparse</li>
    <li>rocmPackages.hsa-amd-aqlprofile-bin</li>
    <li>rocmPackages.llvm.bintools</li>
    <li>rocmPackages.llvm.clang</li>
    <li>rocmPackages.llvm.clang-tools-extra</li>
    <li>rocmPackages.llvm.clang-tools-extra.doc</li>
    <li>rocmPackages.llvm.clang-tools-extra.info</li>
    <li>rocmPackages.llvm.clang-tools-extra.man</li>
    <li>rocmPackages.llvm.clang-unwrapped</li>
    <li>rocmPackages.llvm.clang-unwrapped.doc</li>
    <li>rocmPackages.llvm.clang-unwrapped.info</li>
    <li>rocmPackages.llvm.clang-unwrapped.man</li>
    <li>rocmPackages.llvm.compiler-rt</li>
    <li>rocmPackages.llvm.libc</li>
    <li>rocmPackages.llvm.libc.doc</li>
    <li>rocmPackages.llvm.libcxx</li>
    <li>rocmPackages.llvm.libcxx.doc</li>
    <li>rocmPackages.llvm.libcxxabi</li>
    <li>rocmPackages.llvm.libunwind</li>
    <li>rocmPackages.llvm.libunwind.doc</li>
    <li>rocmPackages.llvm.lld</li>
    <li>rocmPackages.llvm.lld.doc</li>
    <li>rocmPackages.llvm.lldb</li>
    <li>rocmPackages.llvm.lldb.doc</li>
    <li>rocmPackages.llvm.lldb.info</li>
    <li>rocmPackages.llvm.lldb.man</li>
    <li>rocmPackages.llvm.llvm</li>
    <li>rocmPackages.llvm.llvm.doc</li>
    <li>rocmPackages.llvm.llvm.info</li>
    <li>rocmPackages.llvm.llvm.man</li>
    <li>rocmPackages.llvm.mlir</li>
    <li>rocmPackages.llvm.openmp</li>
    <li>rocmPackages.llvm.openmp.doc</li>
    <li>rocmPackages.llvm.openmp.info</li>
    <li>rocmPackages.llvm.openmp.man</li>
    <li>rocmPackages.llvm.polly</li>
    <li>rocmPackages.llvm.polly.doc</li>
    <li>rocmPackages.llvm.polly.info</li>
    <li>rocmPackages.llvm.polly.man</li>
    <li>rocmPackages.llvm.pstl</li>
    <li>rocmPackages.llvm.rocmClangStdenv</li>
    <li>rocmPackages.miopen</li>
    <li>rocmPackages.rccl</li>
    <li>rocmPackages.rocalution</li>
    <li>rocmPackages.rocblas</li>
    <li>rocmPackages.rocdbgapi</li>
    <li>rocmPackages.rocdbgapi.doc</li>
    <li>rocmPackages.rocfft</li>
    <li>rocmPackages.rocgdb</li>
    <li>rocmPackages.rocm-cmake</li>
    <li>rocmPackages.rocm-comgr</li>
    <li>rocmPackages.rocm-core</li>
    <li>rocmPackages.rocm-device-libs</li>
    <li>rocmPackages.rocm-runtime</li>
    <li>rocmPackages.rocm-smi</li>
    <li>rocmPackages.rocm-thunk</li>
    <li>rocmPackages.rocminfo</li>
    <li>rocmPackages.rocmlir</li>
    <li>rocmPackages.rocmlir-rock</li>
    <li>rocmPackages.rocmlir.external</li>
    <li>rocmPackages.rocprim</li>
    <li>rocmPackages.rocprofiler</li>
    <li>rocmPackages.rocr-debug-agent</li>
    <li>rocmPackages.rocsolver</li>
    <li>rocmPackages.rocsparse</li>
    <li>rocmPackages.rocthrust</li>
    <li>rocmPackages.roctracer</li>
    <li>rocmPackages.rocwmma</li>
    <li>rocmPackages.rpp (rocmPackages.rpp-hip)</li>
    <li>rocmPackages.rpp-cpu</li>
    <li>rocmPackages.rpp-opencl</li>
    <li>rocmPackages.tensile</li>
    <li>rocmPackages.tensile.dist</li>
  </ul>
</details>

ulrikstrid reviewed (approved):
I think this looks good now. Landing this would be huge since it takes quite some time to build composable_kernel

mschwaig commented:
> I think this looks good now. Landing this would be huge since it takes quite some time to build composable_kernel

As the author of this PR I think its ready too. 👍 I ran `nixpkgs-review` on it and manually tested if `torchWithRocm` still works.

(This PR is also a prerequisite for merging https://github.com/NixOS/nixpkgs/pull/298388 as is, because that PR simplifies the `rocblas` build in a way that would otherwise exceed the cache size limit as well.)

ulrikstrid commented:
https://github.com/NixOS/nixpkgs/pull/298388#issuecomment-2032791130

This comment is interesting, maybe we can drop those versions that are mentioned? This can be done in a separate PR however so that we can merge this without anything that could be considered controversial sneaking in.

mschwaig commented:
I agree that dropping those ISAs is worth looking into.
I have opened an issue for it, since I do not have the time to open a PR for it in the coming week.

mschwaig commented:
Unfortunately I made a mistake when testing this PR and this segfaults when properly tested with llama.cpp:

```
#0  0x00007fffec3bb452 in amd::Elf::getElfSize(void const*) () from /nix/store/md9m0qlr5qzv1yswqqhx4md4kj8njawf-clr-6.0.2/lib/libamdhip64.so.6
#1  0x00007fffec035025 in hip::CodeObject::ElfSize(void const*) () from /nix/store/md9m0qlr5qzv1yswqqhx4md4kj8njawf-clr-6.0.2/lib/libamdhip64.so.6
#2  0x00007fffec0b043e in hip::FatBinaryInfo::ExtractFatBinaryUsingCOMGR(std::vector<hip::Device*, std::allocator<hip::Device*> > const&) ()
   from /nix/store/md9m0qlr5qzv1yswqqhx4md4kj8njawf-clr-6.0.2/lib/libamdhip64.so.6
#3  0x00007fffec0b1b6e in hip::FatBinaryInfo::ExtractFatBinary(std::vector<hip::Device*, std::allocator<hip::Device*> > const&) () from /nix/store/md9m0qlr5qzv1yswqqhx4md4kj8njawf-clr-6.0.2/lib/libamdhip64.so.6
#4  0x00007fffec038add in hip::StatCO::digestFatBinary(void const*, hip::FatBinaryInfo*&) () from /nix/store/md9m0qlr5qzv1yswqqhx4md4kj8njawf-clr-6.0.2/lib/libamdhip64.so.6
#5  0x00007fffec276a2c in PlatformState::digestFatBinary(void const*, hip::FatBinaryInfo*&) () from /nix/store/md9m0qlr5qzv1yswqqhx4md4kj8njawf-clr-6.0.2/lib/libamdhip64.so.6
#6  0x00007fffec276430 in PlatformState::init() () from /nix/store/md9m0qlr5qzv1yswqqhx4md4kj8njawf-clr-6.0.2/lib/libamdhip64.so.6
#7  0x00007fffec04d0f6 in hip::init(bool*) () from /nix/store/md9m0qlr5qzv1yswqqhx4md4kj8njawf-clr-6.0.2/lib/libamdhip64.so.6
#8  0x00007fffec066dcd in void std::__invoke_impl<void, void (&)(bool*), bool*>(std::__invoke_other, void (&)(bool*), bool*&&) () from /nix/store/md9m0qlr5qzv1yswqqhx4md4kj8njawf-clr-6.0.2/lib/libamdhip64.so.6
#9  0x00007fffec066d9d in std::__invoke_result<void (&)(bool*), bool*>::type std::__invoke<void (&)(bool*), bool*>(void (&)(bool*), bool*&&) ()
   from /nix/store/md9m0qlr5qzv1yswqqhx4md4kj8njawf-clr-6.0.2/lib/libamdhip64.so.6
#10 0x00007fffec066d6c in std::call_once<void (&)(bool*), bool*>(std::once_flag&, void (&)(bool*), bool*&&)::{lambda()#1}::operator()() const ()
   from /nix/store/md9m0qlr5qzv1yswqqhx4md4kj8njawf-clr-6.0.2/lib/libamdhip64.so.6
#11 0x00007fffec066d44 in std::once_flag::_Prepare_execution::_Prepare_execution<std::call_once<void (&)(bool*), bool*>(std::once_flag&, void (&)(bool*), bool*&&)::{lambda()#1}>(void (&)(bool*))::{lambda()#1}::operator()() const () from /nix/store/md9m0qlr5qzv1yswqqhx4md4kj8njawf-clr-6.0.2/lib/libamdhip64.so.6
#12 0x00007fffec066d11 in std::once_flag::_Prepare_execution::_Prepare_execution<std::call_once<void (&)(bool*), bool*>(std::once_flag&, void (&)(bool*), bool*&&)::{lambda()#1}>(void (&)(bool*))::{lambda()#1}::__invoke() () from /nix/store/md9m0qlr5qzv1yswqqhx4md4kj8njawf-clr-6.0.2/lib/libamdhip64.so.6
#13 0x00007fffebbf50df in __pthread_once_slow () from /nix/store/ddwyrxif62r8n6xclvskjyy6szdhvj60-glibc-2.39-5/lib/libc.so.6
#14 0x00007fffec05e5eb in __gthread_once(int*, void (*)()) () from /nix/store/md9m0qlr5qzv1yswqqhx4md4kj8njawf-clr-6.0.2/lib/libamdhip64.so.6
#15 0x00007fffec061fe1 in void std::call_once<void (&)(bool*), bool*>(std::once_flag&, void (&)(bool*), bool*&&) () from /nix/store/md9m0qlr5qzv1yswqqhx4md4kj8njawf-clr-6.0.2/lib/libamdhip64.so.6
#16 0x00007fffec077144 in hipGetDeviceCount () from /nix/store/md9m0qlr5qzv1yswqqhx4md4kj8njawf-clr-6.0.2/lib/libamdhip64.so.6
#17 0x00007ffff6515442 in (anonymous namespace)::TensileHost::TensileHost() () from /nix/store/rydjpmr5awlzznyp4a1r3fhwvyb4gnaj-rocblas-6.0.2/lib/librocblas.so.4
#18 0x00007ffff65098de in (anonymous namespace)::get_library_and_adapter(std::shared_ptr<Tensile::MasterSolutionLibrary<Tensile::ContractionProblem, Tensile::ContractionSolution> >*, std::shared_ptr<hipDeviceProp_tR0600>*, int) () from /nix/store/rydjpmr5awlzznyp4a1r3fhwvyb4gnaj-rocblas-6.0.2/lib/librocblas.so.4
#19 0x00007ffff7e32d56 in ggml_cuda_init() () from /nix/store/9sq0q30g65xyd02yzvixfl39y6wdw5df-llama-cpp-rocm-0.0.0/lib/libllama.so
#20 0x00007ffff7e33399 in ggml_backend_cuda_get_device_count () from /nix/store/9sq0q30g65xyd02yzvixfl39y6wdw5df-llama-cpp-rocm-0.0.0/lib/libllama.so
#21 0x000055555559c31f in _GLOBAL__sub_I_llama_bench.cpp ()
#22 0x00007fffebb8c23e in __libc_start_main_impl () from /nix/store/ddwyrxif62r8n6xclvskjyy6szdhvj60-glibc-2.39-5/lib/libc.so.6
#23 0x0000555555572275 in _start ()
```

mschwaig commented:
`clang-offload-bundler` in `rocmPackages.llvm.openmp` also needs `rocmPackages.llvm.llvm` on it's path to even get there.

@GZGavinZhao do you happen to have a working version of those patches?
I saw in https://github.com/llvm/llvm-project/pull/67162 that you have https://github.com/GZGavinZhao/rocm-llvm-project/commit/c859f20831f33ebfba3fe2f7e8c56877f1b92ba9

ulrikstrid commented:
Should we roll back this or @mschwaig ? Or does someone have time to fix it before 24.05?

GZGavinZhao commented:
> @GZGavinZhao do you happen to have a working version of those the patch for llvm/clang?

I couldn't get LLVM to build with my patches. IIRC it would fail during the test phase, which was sort of expected since I removed the test file changes in the patches to prevent conflicts, but I didn't have time to dig deep into it.

From the stack trace, it seems like the problem is not with LLVM's side, but that the HIP clr runtime and/or ROCm-CompilerSupport doesn't support compressed device binaries yet.

I was able to reproduce this segmentation fault just with the `vectoradd_hip` example file. However, I'd note that upon inspecting the verbose compilation output, it seems like `clang-offload-bundler` compress the device binary _regardless of whether `--offload-compress` was passed_. You can verify this by adding the `--verbose` flag when compiling and you would see this 
```
 "/nix/store/y44kwxkk8pc4wdj0gqhnbdzjm4c7w82w-rocm-llvm-clang-6.0.2/bin/clang-offload-bundler" -type=o -bundle-align=4096 -targets=host-x86_64-unknown-linux,hipv4-amdgcn-amd-amdhsa--gfx1032,hipv4-amdgcn-amd-amdhsa--gfx90c -input=/dev/null -input=/tmp/vectoradd_hip-gfx1032-406a80.out -input=/tmp/vectoradd_hip-gfx90c-2193ac.out -output=/tmp/vectoradd_hip-71eb0a.hipfb -verbose
Compressed bundle format version: 1
Compression method used: zlib
Binary size before compression: 67184 bytes
Binary size after compression: 27668 bytes
Truncated MD5 hash: 0x5296b68059d8fbbe
===-------------------------------------------------------------------------===
                     Timer group for clang offload bundler
===-------------------------------------------------------------------------===
  Total Execution Time: 0.0051 seconds (0.0051 wall clock)

   ---User Time---   --System Time--   --User+System--   ---Wall Time---  --- Name ---
   0.0049 ( 99.4%)   0.0000 (  0.0%)   0.0049 ( 96.8%)   0.0050 ( 96.8%)  Compression time
   0.0000 (  0.6%)   0.0001 (100.0%)   0.0002 (  3.2%)   0.0002 (  3.2%)  Hash calculation time
   0.0050 (100.0%)   0.0001 (100.0%)   0.0051 (100.0%)   0.0051 (100.0%)  Total
```

This is weird because according to llvm/llvm-project#67162, device binary compression should be off by default, but it seems like `clang-offload-bundler` would compress the device binary even if it doesn't receive the `-compress` flag.

mschwaig commented:
Thanks for the minimal test case and all of the additional info.

Compression always being on is due to the wrapper I added here:

https://github.com/NixOS/nixpkgs/blob/772dbad3d41932f08d44aaa63a571d4e26c5d143/pkgs/development/rocm-modules/6/llvm/stage-1/clang-unwrapped.nix#L42-L51

mschwaig commented:
> Should we roll back this @mschwaig ? Or does someone have time to fix it before 24.05?

I don't quite understand the timelines involved, as far as I know this would have to be fixed until tomorrow, to get into the release.

**Until tomorrow (or within the next week even) I don't think this change can be salvaged.**
I could spend more time to look into it starting in a few days, and maybe find and easy fix but the much more likely outcome is  that this is not possible yet, because not all required components support it and we have to look at my previous proposal for working around the size limit again (https://github.com/NixOS/nixpkgs/commit/f199cf1c4eaafd411d9bec764645ff7ae433476c).

We could
* revert the changes in this PR, or
* replace them with what I proposed previously.

I would assume that getting around the Hydra size limit with either solution is something we could backport into the release later on as well (maybe after we have something working on unstable for some time)?
 In that case I think we should just revert this PR.

wegank commented:
The deadline is certainly not tomorrow, as it only applies to breaking changes to [Release Critical Packages](https://nixos.github.io/release-wiki/Release-Critical-Packages.html).

GZGavinZhao commented:
Thanks to a very new PR (4 days ago) llvm/llvm-project#88827, I managed to make compressed device binaries work for `vectoradd_hip.cpp`. Next week I will start cleaning up my work locally, try backporting the commit, and run tests on composable_kernel.

ulrikstrid commented:
@GZGavinZhao please ping me for testing etc
